### PR TITLE
[RISC-V] Fix constant preparing for shuffle stub

### PR DESCRIPTION
**Pull request by t-mustafin**
Part of #84834.

cc @clamp03 @alpencolt @gbalykov

### CODE REVIEW

**Comment by t-mustafin** [comment]
Fixes
```
JIT/Directed/debugging/debuginfo/tester/tester.sh
JIT/Generics/ConstrainedCall/class2_d/class2_d.sh           
JIT/Generics/ConstrainedCall/class2_r/class2_r.sh           
JIT/Generics/Coverage/chaos55915408cs/chaos55915408cs.sh    
JIT/Generics/Coverage/chaos55915408cs_o/chaos55915408cs_o.sh
JIT/Generics/Coverage/chaos56200037cs/chaos56200037cs.sh    
JIT/Generics/Coverage/chaos56200037cs_o/chaos56200037cs_o.sh
JIT/Generics/Coverage/chaos65204782cs/chaos65204782cs.sh    
JIT/Generics/Coverage/chaos65204782cs_o/chaos65204782cs_o.sh
JIT/Intrinsics/TypeIntrinsics_ro/TypeIntrinsics_ro.sh                                                                
JIT/Intrinsics/TypeIntrinsics_r/TypeIntrinsics_r.sh
JIT/Performance/CodeQuality/Roslyn/CscBench/CscBench.sh
JIT/Performance/CodeQuality/Serialization/Deserialize/Deserialize.sh                                                 
JIT/Performance/CodeQuality/Serialization/Serialize/Serialize.sh 
JIT/Performance/CodeQuality/Span/Indexer/Indexer.sh
JIT/Performance/CodeQuality/Span/SpanBench/SpanBench.sh
JIT/Regression/CLR-x86-JIT/V1.2-Beta1/b210352/csharptester/csharptester.sh
JIT/Regression/Dev11/External/Dev11_243742/app/app.sh
```

**Comment by t-mustafin** [comment]
Use algorithm implemented in #86790 allows to decrease address preparing instructions from 7 to 4:
```asm
Before optimization:
   li      a1,63
   slli    a1,a1,0xb
   addi    a1,a1,963
   slli    a1,a1,0xb
   addi    a1,a1,640
   slli    a1,a1,0xa
   addi    a1,a1,848

After optimization:
   lui     a1,0x7ef0e
   addi    a1,a1,-2031 # 0x7ef0d811
   slli    a1,a1,0x7
   addi    a1,a1,88
```

File: src/coreclr/vm/riscv64/stubs.cpp [hunk 1]
```diff
@@ -1009,55 +1009,63 @@ BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDes
...
+
+        if (low11) {
+            EmitSllImm(reg, reg, shiftAccumulator);
+            shiftAccumulator = 0;
+
+            EmitAddImm(reg, reg, low11);
         }
     }
+
+    if (shiftAccumulator) {
```
Comment (alpencolt) [inline src/coreclr/vm/riscv64/stubs.cpp:1066]: Would be good to add notes about `shiftAccumulator`

**Review by clamp03 — APPROVED** [review]
Thank you!

**Review by alpencolt — APPROVED** [review]

**Comment by t-mustafin** [comment]
This PR fixes ~1616 coreclr tests including ~18 tests in JIT directory for Debug build.

With this PR:
```
# Overlay          : /home/jenkins/workspace/dotnet_ubuntu/custom/tests/coreclr_default/coreclr_tests/coreclr-tc/coreroot
# Tests Discovered : 10583
# Passed           : 10287
# Failed           : 252
# Skipped          : 44
```
JIT directory with this PR:
```
# Overlay          : /home/tmustafin/coreclr_tests/coreroot/
# Tests Discovered : 7240
# Passed           : 7134
# Failed           : 83
# Skipped          : 23
```

cc @jkotas please take a look.

File: src/coreclr/vm/riscv64/stubs.cpp [hunk 2]
```diff
@@ -1009,55 +1009,67 @@ BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDes
...
+    UINT32 msb;
+    UINT32 high31;
+
+    BitScanReverse64(&msb, imm);
+
+    if (msb > 30) {
+        high31 = (imm >> (msb - 30)) & 0x7FffFFff;
+    } else {
+        high31 = imm & 0x7FffFFff;
+    }
```
Comment (jkotas) [inline src/coreclr/vm/riscv64/stubs.cpp:1033]:
```suggestion
    if (msb > 30)
    {
        high31 = (imm >> (msb - 30)) & 0x7FffFFff;
    }
    else
    {
        high31 = imm & 0x7FffFFff;
    }
```
Nit: I would be nice to follow consistent formatting with the rest of the file.

File: src/coreclr/vm/riscv64/stubs.cpp [hunk 3]
```diff
@@ -1009,55 +1009,67 @@ BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDes
...
+    // And load remaining part by batches of 11 bits size.
+    INT32 remainingShift = msb - 30;
+
+    // shiftAccumulator usage is an optimization allows to exclude `slli addi` iteration
+    // if immediate bits `low11` for this iteration are zero.
+    UINT32 shiftAccumulator = 0;
+
+    while (remainingShift > 0) {
+        UINT32 shift = remainingShift >= 11 ? 11 : remainingShift % 11;
+        UINT32  mask  = 0x7ff >> (11 - shift);
```
Comment (jkotas) [inline src/coreclr/vm/riscv64/stubs.cpp:1057]:
```suggestion
        UINT32 mask = 0x7ff >> (11 - shift);
```

File: src/coreclr/vm/riscv64/stubs.cpp [hunk 4]
```diff
@@ -1009,55 +1009,67 @@ BOOL GetAnyThunkTarget (T_CONTEXT *pctx, TADDR *pTarget, TADDR *pTargetMethodDes
...
+        remainingShift -= shift;
+        ssize_t low11 = (imm >> remainingShift) & mask;
```
Comment (jkotas) [inline src/coreclr/vm/riscv64/stubs.cpp:1059]:
```suggestion
        UINT32 low11 = (imm >> remainingShift) & mask;
```
Everything else in this method is UINT32. Is there a reason for `ssize_t` here?

Comment (t-mustafin) [reply src/coreclr/vm/riscv64/stubs.cpp]: Thank you!

File: src/coreclr/vm/riscv64/stubs.cpp (hunk 2 above)
Comment (t-mustafin) [reply src/coreclr/vm/riscv64/stubs.cpp]: Thanks! In this file I found both types of bracket style.

**Review by jkotas — APPROVED** [review]
Thank you!

**Pull request merged**